### PR TITLE
35 single pas DOF

### PR DESCRIPTION
**Pull request by simonharrisco**
- Remove old  dof textures and add new one in post-processing.js
- Add new dof pass writing to new texture
- Add dof.frag.js
- Make final pass color input depend on if dof is active or not
- Change DOF params:
  - Keep dofDepth
  - Add dofScale
  - Remove dofIterations, dofRange, dofRadius
- Update examples/post-processing and examples/window-resize to take new params
- Updated examples/post-processing to tidy frame()

Some things that probably require thought going forward:
- param naming: dofScale -> the higher this is the smaller the area thats in focus. I dont think this is a good name and isnt too desciptive so open too suggestions
- shader code not been changed since last project so probably still resolution dependant

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1325,6 +1313,12 @@ Renderer.prototype.draw = function() {
             uFogDensity: postProcessingCmp.fogDensity,
             uSunPosition: postProcessingCmp.sunPosition,
             uOutputEncoding: ctx.Encoding.Gamma,
+            uOverlay:
```
Comment (dmnsgn) [inline index.js]: `postProcessingCmp && postProcessingCmp.enabled ` don't need to be checked, you are already in this condition

File: shaders/post-processing/dof.frag.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+uniform float uNear;
+uniform float uFocusPoint;
+uniform float uFocusScale;
+
+const float GOLDEN_ANGLE = 2.39996323; 
+const float MAX_BLUR_SIZE = 20.0; 
+const float RAD_SCALE = 0.5; // Smaller = nicer blur, larger = faster
+
+
+float readDepth(const in sampler2D depthMap, const in vec2 coord, const in float near, const in float far) {
```
Comment (dmnsgn) [inline shaders/post-processing/dof.frag.js]: Use the existing chunk here: https://github.com/pex-gl/pex-renderer/blob/master/shaders/chunks/depth-read.glsl.js

File: shaders/post-processing/bilateral-blur.frag.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ uniform vec2 direction;
 
 uniform float sharpness;
 uniform float uDOFDepth;
-uniform float uDOFRange;
+uniform float udofScale;
```
Comment (dmnsgn) [inline shaders/post-processing/bilateral-blur.frag.js]: Shall we keep the same naming pattern here? `uDOFScale`

**Comment by simonharrisco** [comment]
Bilateral blur is no longer used for dof so i think it can be removed from the shader now

**Comment by vorg** [comment]
@simonharrisco what about `dofFocus`? Also why don't we do the usual and check Three/Babylon/Unity param names?

**Comment by simonharrisco** [comment]
Unity called it 'Focal Length' https://docs.unity3d.com/Packages/com.unity.postprocessing@2.1/manual/Depth-of-Field.html

Babylon doesnt have an option to change it, but lets you do lots of other things:
https://doc.babylonjs.com/how_to/using_depth-of-field_and_other_lens_effects

**Comment by vorg** [comment]
### Babylon
https://doc.babylonjs.com/how_to/using_depth-of-field_and_other_lens_effects
- `dof_focus_distance` - The distance at which focus is set. Everything out of focus will be blurred. Note that a focus distance above a few hundred units gives you infinity focus.
- `dof_aperture` - Default: 1. How quickly things outside of focus will become blurred. Set high to give a ‘miniature’ effect (a.k.a. tilt shift).


### Unity
https://docs.unity3d.com/Packages/com.unity.postprocessing@2.1/manual/Depth-of-Field.html
- `Focus Distance` - Set the distance to the point of focus.
- `Aperture` - Set the ratio of the aperture (known as f-stop or f-number). The smaller the value is, the shallower the depth of field is.
- `Focal Length` - Set the distance between the lens and the film. The larger the value is, the shallower the depth of field is. (edited) 

### UE4 (with interactive examples!)
https://docs.unrealengine.com/en-US/Engine/Rendering/PostProcessEffects/DepthOfField/CinematicDOFMethods/index.html
- `Aperture` - The Aperture is the diameter of the opening through which light passes. The Aperture is defined by Focal Length divided by an F-stop number.

**Comment by vorg** [comment]
Related https://github.com/pex-gl/pex-renderer/issues/232

**Comment by vorg** [comment]
@simonharrisco so if we go for aperture what is actual value? 1.4 ~ 16? And we clamp/remap it somehow so it looks good?

**Comment by vorg** [comment]
@simonharrisco could we add DOF only example? No orbiter, 3 aperture views, focus distance slider, resolution check boxes (so we can test resolution independence)?

IMG_8050

**Pull request merged**